### ISSUE TITLE
[Bug] TONS of Claim Blocks Being Given To Random(?) Players For No Known Reason

### ISSUE THREAD

**Issue by PiggiesGoSqueal** [GriefPrevention/GriefPrevention#1193]
### What happened:
Claim blocks are all messed up.
Seemingly random players are getting absurd numbers of claim blocks for seemingly no reason.

I first noticed this when I joined my server on my player account- which was fairly new and thus should've only had around the starter amount (which is ~2,500). But it had some ridiculous number of claim blocks. I hoped this was a one-time bizarre issue so I reset his claim blocks if I recall correctly. Today a player reported the issue too. This time the player had 31,700 claim blocks! Realistically they should've had around 10k or less.

I ended up checking my player account's PlayerData file for GriefPrevention after this player's bug report. It showed:
100000
18703

Even though holding a golden shovel only shows "You may claim up to 291 more blocks."

Then I checked the player's file who reported it today:
15228
10199

However when they do "/buyclaimblocks 1" they are told they have 31,700!

What is going on??

There are 4 ways to get claim blocks on my server:
1. Donor ranks (Although my player acc didn't have a donor rank and the player who reported the issue today only received 5k from their donor rank. Also I double checked the commands ran when the player donates and it is set to give it to that specific player only.)
2. Hourly (50 claim blocks per hour)
3. The starter claim blocks everyone gets at beginning (~2,500)
4. /buyclaimblocks ($2/block) 

### What was expected:
I expected all players to have normal amounts of claim blocks.

### Steps to reproduce:
As I have no idea what is causing the problem I have no steps to reproduce.

### Server and GriefPrevention version:
___This can be found by running `/version` and `/version GriefPrevention` on your server.___
Paste of `/version`: 
[23:40:53 INFO]: --------------------------------------------------
[23:40:53 INFO]: CMI: 8.7.12.7 MySQL
[23:40:53 INFO]: Server: Paper(346) 1.16.4-R0.1-SNAPSHOT
[23:40:53 INFO]: CMI economy: True Vault: 1.7.3-b CMI Chat: True 
[23:40:53 INFO]: Modules -> 45 enabled 8 disabled: spawnerProximity, cuffed, disabledEnchants, worldLimitsElytra, votifier, skin, warmup, vanish
[23:40:53 INFO]: --------------------------------------------------

Paste of '/ver':
Paper version git-Paper-346 (MC: 1.16.4) (Implementing API version 1.16.4-R0.1-SNAPSHOT)

Paste of `/version GriefPrevention`: 
Doesn't work but latest.log says:
[GriefPrevention] Loading GriefPrevention v16.16.0

### Stack trace/error or server log
latest.log:
https://gist.github.com/PiggiesGoSqueal/693caf49386c036dc40787b6360fbafb

### GriefPrevention config.yml

```yml

# Default values are perfect for most servers.  If you want to customize and have a question, look for the answer here first: http://dev.bukkit.org/bukkit-plugins/grief-prevention/pages/setup-and-configuration/
GriefPrevention:
  SeaLevelOverrides:
    world: -1
    world_nether: -1
    world_the_end: -1
    rw: -1
    spawn: -1
    misc: -1
  Claims:
    Mode:
      world_the_end: Survival
      rw: Disabled
      world_nether: Survival
      spawn: Disabled
      world: Survival
      misc: Survival
    PreventGlobalMonsterEggs: true
    PreventTheft: true
    ProtectCreatures: true
    PreventButtonsSwitches: true
    LockWoodenDoors: true
    LockTrapDoors: true
    LockFenceGates: true
    EnderPearlsRequireAccessTrust: true
    RaidTriggersRequireBuildTrust: true
    ProtectHorses: false
    ProtectDonkeys: false
    ProtectLlamas: false
    InitialBlocks: 2500
    Claim Blocks Accrued Per Hour:
      Default: 50
    Max Accrued Claim Blocks:
      Default: 100000
    Accrued Idle Threshold: 0
    AccruedIdlePercent: 0
    AbandonReturnRatio: 1.0
    AutomaticNewPlayerClaimsRadius: -1
    AutomaticNewPlayerClaimsRadiusMinimum: 0
    ExtendIntoGroundDistance: 256
    MinimumWidth: 1
    MinimumArea: 100
    MaximumDepth: 0
    InvestigationTool: FEATHER
    ModificationTool: GOLDEN_SHOVEL
    Expiration:
      ChestClaimDays: 10000
      UnusedClaimDays: 10000
      AllClaims:
        DaysInactive: 90
        ExceptWhenOwnerHasTotalClaimBlocks: 10000000
        ExceptWhenOwnerHasBonusClaimBlocks: 50000000
      AutomaticNatureRestoration:
        SurvivalWorlds: false
    AllowTrappedInAdminClaims: false
    MaximumNumberOfClaimsPerPlayer: 0
    CreationRequiresWorldGuardBuildPermission: true
    VillagerTradingRequiresPermission: true
    CommandsRequiringAccessTrust: /sethome
    DeliverManuals: true
    ManualDeliveryDelaySeconds: 30
    RavagersBreakBlocks: true
    FireSpreadsInClaims: true
    FireDamagesInClaims: true
    LecternReadingRequiresAccessTrust: true
  Spam:
    Enabled: false
    LoginCooldownSeconds: 60
    LoginLogoutNotificationsPerMinute: 5
    ChatSlashCommands: /me;/global;/local
    WhisperSlashCommands: /tell;/pm;/r;/whisper;/msg
    WarningMessage: Please reduce your noise level.  Spammers will be banned.
    BanOffenders: false
    BanMessage: Banned for spam.
    AllowedIpAddresses: 1.2.3.4; 5.6.7.8
    DeathMessageCooldownSeconds: 120
    Logout Message Delay In Seconds: 0
  PvP:
    RulesEnabledInWorld:
      world: true
      world_nether: true
      world_the_end: true
      rw: true
      spawn: true
      misc: true
    ProtectFreshSpawns: false
    PunishLogout: false
    CombatTimeoutSeconds: 0
    AllowCombatItemDrop: true
    BlockedSlashCommands: /fly;/xpfly;/mc;/enderchest;/echest;/ec;/ender;/bp small
      1;/bp medium 1;/bp large 1;/bp large 2;/tpa;/tpaccept;/tpahere
    ProtectPlayersInLandClaims:
      PlayerOwnedClaims: false
      AdministrativeClaims: true
      AdministrativeSubdivisions: true
    AllowLavaDumpingNearOtherPlayers:
      PvPWorlds: false
      NonPvPWorlds: false
    AllowFlintAndSteelNearOtherPlayers:
      PvPWorlds: false
      NonPvPWorlds: false
    ProtectPetsOutsideLandClaims: false
  Economy:
    ClaimBlocksMaxBonus: 0
    ClaimBlocksPurchaseCost: 2.0
    ClaimBlocksSellValue: 0.0
  ProtectItemsDroppedOnDeath:
    PvPWorlds: false
    NonPvPWorlds: false
  BlockLandClaimExplosions: true
  BlockSurfaceCreeperExplosions: true
  BlockSurfaceOtherExplosions: false
  LimitSkyTrees: true
  LimitTreeGrowth: false
  PistonMovement: EVERYWHERE_SIMPLE
  PistonExplosionSound: true
  FireSpreads: true
  FireDestroys: true
  AdminsGetWhispers: false
  AdminsGetSignNotifications: false
  SmartBan: false
  Mute New Players Using Banned Words: false
  MaxPlayersPerIpAddress: 3
  Siege:
    Worlds: []
    BreakableBlocks:
    - DIRT
    - GRASS_BLOCK
    - GRASS
    - FERN
    - DEAD_BUSH
    - COBBLESTONE
    - GRAVEL
    - SAND
    - GLASS
    - GLASS_PANE
    - OAK_PLANKS
    - SPRUCE_PLANKS
    - BIRCH_PLANKS
    - JUNGLE_PLANKS
    - ACACIA_PLANKS
    - DARK_OAK_PLANKS
    - WHITE_WOOL
    - ORANGE_WOOL
    - MAGENTA_WOOL
    - LIGHT_BLUE_WOOL
    - YELLOW_WOOL
    - LIME_WOOL
    - PINK_WOOL
    - GRAY_WOOL
    - LIGHT_GRAY_WOOL
    - CYAN_WOOL
    - PURPLE_WOOL
    - BLUE_WOOL
    - BROWN_WOOL
    - GREEN_WOOL
    - RED_WOOL
    - BLACK_WOOL
    - SNOW
    DoorsOpenDelayInSeconds: 300
    CooldownEndInMinutes: 60
  EndermenMoveBlocks: false
  SilverfishBreakBlocks: true
  CreaturesTrampleCrops: false
  RabbitsEatCrops: true
  HardModeZombiesBreakDoors: false
  Database:
    URL: ''
    UserName: ''
    Password: ''
  UseBanCommand: false
  BanCommandPattern: ban %name% %reason%
  Advanced:
    fixNegativeClaimblockAmounts: true
    ClaimExpirationCheckRate: 90
    OfflinePlayer_cache_days: 90
  Abridged Logs:
    Days To Keep: 7
    Included Entry Types:
      Social Activity: false
      Suspicious Activity: true
      Administrative Activity: false
      Debug: false
      Muted Chat Messages: false
```

### Plugin list (if applicable):
Plugins (49): ActionAnnouncer, AsyncWorldEdit, AutomaticInventory, BeaconEffects*, BlockLocker, BossShopPro, Chunkmaster, ClaimslistClassifier*, CMI, CMIPaperLib, CoreProtect, CraftingStore, CrateReloaded, DiscordSRV, EntityTrackerFixer, ezAuctions, eZProtector, FancyBags, Geyser-Spigot, GriefPrevention, GriefPreventionFlags, IllegalStack, InfiniteJump, LibsDisguises, LuckPerms, mcMMO, MobFarmManager, MorphRedeem, Multiverse-Core, MyCommand, MysqlEconomyBank, PlayerParticles, PremiumVanish, ProtocolLib, Quests, QuickShop, RecipeControl, ResourceWorld, SAML, ShopGUIPlus, SignShop, Vault, VillagerOptimiser, VoidChest, Votifier, VotingPlugin, WorldEdit, WorldGuard, XRayDetector*

### Debugging steps attempted
As the issue is inconsistent I cannot really debug it with my current knowledge?

I will note that by checking a totally random file in PlayerData folder I can confirm some (or most?) files are normal:
2500
0

- [ ] I attempted running only GriefPrevention on the server
  - [ ] Issue still occurs even if GriefPrevention is the only plugin running
- [ ] I attempted testing for the issue on a new server
  - [ ] Issue still occurs even after testing on a new server
- [ ] I attempted testing for the issue with the vanilla client (no mods)
  - [ ] Issue still occurs even after testing with the vanilla Minecraft client

Help would be greatly appreciated. Thank you!

**Comment by RoboMWM** [comment]
enable debug logging in config. This prints everything regarding accruals.

**Comment by RoboMWM** [comment]
It sounds like tho that negative claimblock totals are being corrected. Do you have situations that cause players to have negative claimblock totals (i.e. having more claims than claimblocks)?

**Comment by PiggiesGoSqueal** [comment]
> 
> 
> enable debug logging in config. This prints everything regarding accruals.

How long should I leave it enabled & will it impact performance?
I have just enabled it. 

> 
> 
> It sounds like tho that negative claimblock totals are being corrected. Do you have situations that cause players to have negative claimblock totals (i.e. having more claims than claimblocks)?

Not that I know of. Also I just tested & confirmed that claim blocks are reduced when player claims and all claim blocks return when abandoning a claim (with no extras).

Although the config states:
  Advanced:
    **fixNegativeClaimblockAmounts: true**
    ClaimExpirationCheckRate: 90
    OfflinePlayer_cache_days: 90
Is it supposed to be this?

**Comment by RoboMWM** [comment]
no, debug logging never impacts performance.

**Comment by PiggiesGoSqueal** [comment]
> 
> 
> no, debug logging never impacts performance.

Sorry for late reply! Winter break ended so life has been/is busy.

I've provided a compressed file of last 8 days of debug logs. Hopefully that will be enough to determine the issue.
[GP-Log-Files.zip](https://github.com/TechFortress/GriefPrevention/files/5859575/GP-Log-Files.zip)
Also just a side note in case it matters, today I upgraded to Paper 1.16.5 (from 1.16.4) so the log for 1/22/2021 may or may not be for that Paper version, idk.

Edit: VirusTotal scan result for that download link:
https://www.virustotal.com/gui/url/a263d2c65baf5155f6ce4643d8916dbcf95b3d629ab4ca3efa8552bc1981a06e/detection

Edit 2: based off most recent log file it looks like it periodically gives several players 8 claim blocks.
I.e. "Delivering 8 blocks to NotUneay"

For the relevant part of config I have:
    InitialBlocks: 2500
    Claim Blocks Accrued Per Hour:
      Default: 50
    Max Accrued Claim Blocks:
      Default: 10000
    Accrued Idle Threshold: 0
    AccruedIdlePercent: 0
    AbandonReturnRatio: 1.0
   

Latest debug log file: https://gist.github.com/PiggiesGoSqueal/9b7f05b9618d7432afa480ecb7d46848
Here are the timestamps for an example:
Line 593: 04:47 Delivering 8 blocks to Sausagis
Line 605: 04:57 Delivering 8 blocks to Sausagis

Assuming the timestamp is HH:MM format then 8 claim blocks per 10 min for 1 hour = 48 claim blocks per hour. So I suppose it makes sense for the config setup & wouldn't be enough to reach the huge number of claim blocks some players have including my newer player acc.

**Comment by RoboMWM** [comment]
Has the issue (or similar new issues) occurred during this timeframe?

In both files it appears the accrued value never exceeds your configured amount. So I don't think it's accruals.

I don't immediately see any plugins you have that can transfer claims, so I wonder if an admin perhaps facilitated a /transferclaim  with/to that player at one point? The `/transferclaim` command doesn't transfer claimblocks, and might be the thing triggering the extra blocks to be added in that one scenario.

**Comment by PiggiesGoSqueal** [comment]
> 
> 
> Has the issue (or similar new issues) occurred during this timeframe?
> 
> In both files it appears the accrued value never exceeds your configured amount. So I don't think it's accruals.
> 
> I don't immediately see any plugins you have that can transfer claims, so I wonder if an admin perhaps facilitated a /transferclaim with/to that player at one point? The `/transferclaim` command doesn't transfer claimblocks, and might be the thing triggering the extra blocks to be added in that one scenario.

Sorry for very late reply kinda forgot about this. There haven't been further reports of it but I seriously doubt it was solved as nothing has been changed for the plugin. I'm the only admin with /transferclaim permissions and I've never used that command before. 

I appreciate your patience btw with my very very slow replies. :)

**Comment by PiggiesGoSqueal** [comment]
Well, I FINALLY figured out the issue. 

When I make a new server map I'll sometimes take files from the old map and edit them to my liking to save time. I somehow managed to make the very dumb mistake of transferring player data as well. So all claims from the old map existed on the new map. And if a player did /abandonallclaims then all of those claim blocks would be usable to them- hence why it seemed like players would randomly get a ton.

Eventually a player pointed out they had claims that they never made (& were outside the world border). I compared them between the old map & new map files and they were the same. 

I removed the claims, although to my knowledge there is no efficient way to remove the claim blocks from the old map & keep the claim blocks from the new map. Rip.
